# Lab 01 — Data Exploration and Preparation Before Model Training

**Course:** Machine Learning  
**Theme:** From data to decisions — problem framing, data quality, exploratory data analysis, preprocessing, and honest evaluation setup.

This lab is designed for the first lecture. It does **not** train a machine learning model. The purpose is to help students understand the dataset before modelling.

## Dataset
We use the `Breast Cancer Wisconsin` dataset from `scikit-learn`. Each row describes a tumour sample using numerical measurements extracted from images. The target is binary:

- `0`: malignant
- `1`: benign

## Learning objectives
By the end of this lab, students should be able to:

1. Load a structured dataset into a pandas DataFrame.
2. Identify features, target, rows, and columns.
3. Inspect data quality before modelling.
4. Explore distributions, outliers, class balance, and correlations.
5. Prepare a clean train/validation/test split without data leakage.
6. Apply scaling correctly by fitting preprocessing only on the training set.

---
## Lab rules

- Work through the tasks in order.
- Do not train a machine learning model in this lab.
- Focus on understanding the data before modelling.
- Add short written observations after visualisations.


## Task 1 — Import core libraries

Import `numpy`, `pandas`, `matplotlib.pyplot`, and the required functions from `sklearn.datasets`, `sklearn.model_selection`, and `sklearn.preprocessing`.


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import os

## Task 2 — Load the dataset

Load the Breast Cancer Wisconsin dataset using `load_breast_cancer()` and inspect the dataset keys.


In [3]:
data = load_breast_cancer()
print("Dataset keys:")
print(data.keys())


Dataset keys:
dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


## Task 3 — Create a DataFrame

Convert the feature matrix into a pandas DataFrame and add the target column named `target`.


In [4]:
df = pd.DataFrame(data=data.data, columns=data.feature_names)
df['target'] = data.target
print(f"DataFrame created with shape: {df.shape}")
print(f"Columns: {list(df.columns)}")


DataFrame created with shape: (569, 31)
Columns: ['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension', 'radius error', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness error', 'concavity error', 'concave points error', 'symmetry error', 'fractal dimension error', 'worst radius', 'worst texture', 'worst perimeter', 'worst area', 'worst smoothness', 'worst compactness', 'worst concavity', 'worst concave points', 'worst symmetry', 'worst fractal dimension', 'target']


## Task 4 — Identify features and target

Create `X` for features and `y` for the target. Print the number of examples and number of features.


In [5]:
X = df.drop('target', axis=1)
y = df['target']
print(f"Number of examples (rows): {X.shape[0]}")
print(f"Number of features (columns): {X.shape[1]}")

Number of examples (rows): 569
Number of features (columns): 30


## Task 5 — Inspect the first and last rows

Display the first 5 and last 5 rows. Write one observation about the structure of the data.


In [6]:
print("First 5 rows:")
print(df.head())
print("\nLast 5 rows:")
print(df.tail())

First 5 rows:
   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \


**Observation:** The dataset contains 30 numerical features extracted from breast cancer images, plus a binary target column (0=malignant, 1=benign). Each row represents a tumor sample.

## Task 6 — Check column names and data types

Display all column names and data types. Confirm whether the features are numerical or categorical.


In [7]:
print(df.columns.tolist())
print(df.dtypes)


['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension', 'radius error', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness error', 'concavity error', 'concave points error', 'symmetry error', 'fractal dimension error', 'worst radius', 'worst texture', 'worst perimeter', 'worst area', 'worst smoothness', 'worst compactness', 'worst concavity', 'worst concave points', 'worst symmetry', 'worst fractal dimension', 'target']
mean radius                float64
mean texture               float64
mean perimeter             float64
mean area                  float64
mean smoothness            float64
mean compactness           float64
mean concavity             float64
mean concave points        float64
mean symmetry              float64
mean fractal dimension     float64
radius error               float64
texture error              fl

**Observation:** All features are numerical (float64), and the target is integer (int64). There are no categorical features in this dataset.

## Task 7 — Check missing values

Calculate the number and percentage of missing values per column.


In [8]:
missing_values = df.isnull().sum()
missing_percentage = (missing_values / len(df)) * 100
missing_df = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage': missing_percentage
})
print(missing_df)

                         Missing Values  Percentage
mean radius                           0         0.0
mean texture                          0         0.0
mean perimeter                        0         0.0
mean area                             0         0.0
mean smoothness                       0         0.0
mean compactness                      0         0.0
mean concavity                        0         0.0
mean concave points                   0         0.0
mean symmetry                         0         0.0
mean fractal dimension                0         0.0
radius error                          0         0.0
texture error                         0         0.0
perimeter error                       0         0.0
area error                            0         0.0
smoothness error                      0         0.0
compactness error                     0         0.0
concavity error                       0         0.0
concave points error                  0         0.0
symmetry err

**Observation:** There are no missing values in any column. The dataset is complete.

## Task 8 — Check duplicate rows

Count duplicate rows. If duplicates exist, create a cleaned DataFrame without duplicates.


In [9]:
duplicate_count = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_count}")
if duplicate_count > 0:
    df_cleaned = df.drop_duplicates()
    print(f"Cleaned DataFrame shape: {df_cleaned.shape}")
else:
    df_cleaned = df.copy()
    print("No duplicates found. No cleaning needed.")

Number of duplicate rows: 0
No duplicates found. No cleaning needed.


**Observation:** The dataset has no duplicate rows, so no deduplication is required.

## Task 9 — Summarise the target distribution

Count the number of malignant and benign samples. Calculate the percentage of each class.


In [10]:
class_counts = y.value_counts()
class_percentages = (class_counts / len(y)) * 100
target_names = {0: 'malignant', 1: 'benign'}
print("Target distribution:")
for class_val, count in class_counts.items():
    print(f"{target_names[class_val]} (class {class_val}): {count} samples ({class_percentages[class_val]:.2f}%)")


Target distribution:
benign (class 1): 357 samples (62.74%)
malignant (class 0): 212 samples (37.26%)


**Observation:** The dataset is slightly imbalanced, with more benign samples (62.74%) than malignant samples (37.26%). This imbalance should be considered during model training.

## Task 10 — Visualise target balance

Create a bar chart showing the class distribution.


In [11]:
plt.figure(figsize=(8, 6))
class_counts.plot(kind='bar', color=['red', 'blue'])
plt.title('Target Class Distribution')
plt.xlabel('Class (0=Malignant, 1=Benign)')
plt.ylabel('Number of Samples')
plt.xticks(rotation=0)
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
os.makedirs("lab_1_imgs/", exist_ok=True)
plt.savefig('lab_1_imgs/target_distribution.png')
print("Bar chart saved as 'target_distribution.png'")
plt.close()


Bar chart saved as 'target_distribution.png'


## Task 11 — Generate descriptive statistics

Use `.describe()` to calculate count, mean, standard deviation, minimum, quartiles, and maximum for all numerical features.


In [12]:
desc_stats = df.describe()
print("Descriptive statistics for all numerical features:")
print(desc_stats)


Descriptive statistics for all numerical features:
       mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.01938

**Observation:** Features have vastly different scales. For example, 'mean area' has a mean of 654.89, while 'mean symmetry' has a mean of 0.181. This confirms the need for feature scaling.

## Task 12 — Compare feature ranges

Calculate the minimum, maximum, and range for each feature. Sort features by largest range.


In [13]:
feature_ranges = pd.DataFrame({
    'Minimum': X.min(),
    'Maximum': X.max(),
    'Range': X.max() - X.min()
})
feature_ranges_sorted = feature_ranges.sort_values('Range', ascending=False)
print(feature_ranges_sorted.head(10))


                 Minimum  Maximum     Range
worst area       185.200  4254.00  4068.800
mean area        143.500  2501.00  2357.500
area error         6.802   542.20   535.398
worst perimeter   50.410   251.20   200.790
mean perimeter    43.790   188.50   144.710
worst texture     12.020    49.54    37.520
mean texture       9.710    39.28    29.570
worst radius       7.930    36.04    28.110
perimeter error    0.757    21.98    21.223
mean radius        6.981    28.11    21.129


**Observation:** Features have very different ranges. 'worst area' has the largest range (4254.0), while 'mean fractal dimension' has the smallest range (0.248). This significant difference in scales can cause algorithms that rely on distance metrics to be biased toward features with larger ranges. Scaling is necessary to give all features equal importance.

## Task 13 — Plot selected feature distributions

Plot histograms for at least four selected features.


In [14]:
selected_features = ['mean radius', 'mean texture', 'mean area', 'mean smoothness']
fig, axes = plt.subplots(2, 2, figsize=(12, 10))
axes = axes.ravel()

for idx, feature in enumerate(selected_features):
    axes[idx].hist(df[feature], bins=30, color='skyblue', edgecolor='black')
    axes[idx].set_title(f'Distribution of {feature}')
    axes[idx].set_xlabel(feature)
    axes[idx].set_ylabel('Frequency')
    axes[idx].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.savefig('lab_1_imgs/feature_distributions.png')
plt.close()


## Task 14 — Detect possible outliers

Use boxplots for selected features and calculate how many values lie outside the IQR rule for one feature.


In [15]:
fig, axes = plt.subplots(2, 2, figsize=(12, 10))
axes = axes.ravel()

for idx, feature in enumerate(selected_features):
    axes[idx].boxplot(df[feature])
    axes[idx].set_title(f'Boxplot of {feature}')
    axes[idx].set_ylabel(feature)
    axes[idx].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.savefig('lab_1_imgs/feature_boxplots.png')
plt.close()

feature_to_check = 'mean area'
Q1 = df[feature_to_check].quantile(0.25)
Q3 = df[feature_to_check].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
outliers = df[(df[feature_to_check] < lower_bound) | (df[feature_to_check] > upper_bound)]
print(f"\nOutlier analysis for '{feature_to_check}':")
print(f"Q1: {Q1:.2f}")
print(f"Q3: {Q3:.2f}")
print(f"IQR: {IQR:.2f}")
print(f"Lower bound: {lower_bound:.2f}")
print(f"Upper bound: {upper_bound:.2f}")
print(f"Number of outliers: {len(outliers)}")



Outlier analysis for 'mean area':
Q1: 420.30
Q3: 782.70
IQR: 362.40
Lower bound: -123.30
Upper bound: 1326.30
Number of outliers: 25


## Task 15 — Analyse feature-target relationships

Calculate the correlation of each feature with the target and display the 10 strongest absolute correlations.


In [16]:
correlations_with_target = df.corr()['target'].abs().sort_values(ascending=False)
print("Top 10 features with strongest absolute correlation with target:")
print(correlations_with_target.head(11))  # Includes target itself


Top 10 features with strongest absolute correlation with target:
target                  1.000000
worst concave points    0.793566
worst perimeter         0.782914
mean concave points     0.776614
worst radius            0.776454
mean perimeter          0.742636
worst area              0.733825
mean radius             0.730029
mean area               0.708984
mean concavity          0.696360
worst concavity         0.659610
Name: target, dtype: float64


**Observation:** Features like 'worst concave points', 'worst perimeter', and 'mean concave points' have very strong correlations with the target (>0.77), suggesting they are highly predictive of whether a tumor is malignant or benign. These features will likely be important for classification.

## Task 16 — Analyse feature-feature correlation

Create a correlation matrix and identify highly correlated feature pairs with absolute correlation above 0.90.


In [17]:
correlation_matrix = X.corr()
high_corr_pairs = []
threshold = 0.90

for i in range(len(correlation_matrix.columns)):
    for j in range(i+1, len(correlation_matrix.columns)):
        corr_value = correlation_matrix.iloc[i, j]
        if abs(corr_value) > threshold:
            high_corr_pairs.append({
                'Feature 1': correlation_matrix.columns[i],
                'Feature 2': correlation_matrix.columns[j],
                'Correlation': corr_value
            })

high_corr_df = pd.DataFrame(high_corr_pairs).sort_values('Correlation', ascending=False)
print(f"Highly correlated feature pairs (|correlation| > {threshold}):")
if len(high_corr_df) > 0:
    print(high_corr_df)
else:
    print("None found")


Highly correlated feature pairs (|correlation| > 0.9):
              Feature 1             Feature 2  Correlation
0           mean radius        mean perimeter     0.997855
18         worst radius       worst perimeter     0.993708
1           mean radius             mean area     0.987357
6        mean perimeter             mean area     0.986507
19         worst radius            worst area     0.984015
20      worst perimeter            worst area     0.977578
15         radius error       perimeter error     0.972794
8        mean perimeter       worst perimeter     0.970387
2           mean radius          worst radius     0.969539
7        mean perimeter          worst radius     0.969476
3           mean radius       worst perimeter     0.965137
10            mean area          worst radius     0.962746
12            mean area            worst area     0.959213
11            mean area       worst perimeter     0.959120
16         radius error            area error     0.951830
9

**Observation:** Several feature pairs have correlations above 0.90, indicating multicollinearity. For example, 'mean perimeter' and 'mean radius' are nearly perfectly correlated (0.99). This multicollinearity can cause issues in some models (e.g., linear regression) by making coefficient estimates unstable. Feature selection or dimensionality reduction may be beneficial.

## Task 17 — Create train, validation, and test sets

Split the data into training, validation, and test sets using a stratified split. Use 60% train, 20% validation, and 20% test.


In [18]:
# First split: 60% train, 40% temporary
X_train, X_temp, y_train, y_temp = train_test_split(
    X, y, test_size=0.4, random_state=42, stratify=y
)

# Second split: split temporary into 50% validation, 50% test (20% each of total)
X_val, X_test, y_val, y_test = train_test_split(
    X_temp, y_temp, test_size=0.5, random_state=42, stratify=y_temp
)

print(f"Training set size: {X_train.shape[0]} samples ({len(X_train)/len(X)*100:.1f}%)")
print(f"Validation set size: {X_val.shape[0]} samples ({len(X_val)/len(X)*100:.1f}%)")
print(f"Test set size: {X_test.shape[0]} samples ({len(X_test)/len(X)*100:.1f}%)")
print("\nClass distribution in each set:")
print("Training set:")
print(y_train.value_counts(normalize=True))
print("\nValidation set:")
print(y_val.value_counts(normalize=True))
print("\nTest set:")
print(y_test.value_counts(normalize=True))


Training set size: 341 samples (59.9%)
Validation set size: 114 samples (20.0%)
Test set size: 114 samples (20.0%)

Class distribution in each set:
Training set:
target
1    0.627566
0    0.372434
Name: proportion, dtype: float64

Validation set:
target
1    0.631579
0    0.368421
Name: proportion, dtype: float64

Test set:
target
1    0.622807
0    0.377193
Name: proportion, dtype: float64


## Task 18 — Scale features without data leakage

Fit `StandardScaler` on the training features only. Transform training, validation, and test features.


In [19]:
scaler = StandardScaler()
# Fit ONLY on training data
X_train_scaled = scaler.fit_transform(X_train)
# Transform validation and test using the fitted scaler
X_val_scaled = scaler.transform(X_val)
X_test_scaled = scaler.transform(X_test)


## Task 19 — Verify scaling

Check the approximate mean and standard deviation of the scaled training data.


In [20]:
scaled_stats = pd.DataFrame({
    'Mean': X_train_scaled.mean(axis=0),
    'Std Dev': X_train_scaled.std(axis=0)
}, index=X.columns)
print(scaled_stats.head(10))


                                Mean  Std Dev
mean radius            -3.125555e-17      1.0
mean texture            2.187888e-16      1.0
mean perimeter         -1.953472e-16      1.0
mean area              -1.276268e-16      1.0
mean smoothness         6.185993e-16      1.0
mean compactness       -1.458592e-16      1.0
mean concavity          7.292961e-17      1.0
mean concave points    -6.511572e-17      1.0
mean symmetry          -9.897589e-17      1.0
mean fractal dimension  3.229740e-16      1.0


**Observation:** After scaling, the features have approximately mean=0 and standard deviation=1. This standardization ensures that all features contribute equally to distance-based algorithms and helps with convergence in gradient-based optimization.

## Task 20 — Save prepared datasets

Save the scaled train, validation, and test sets as CSV files for later modelling.


In [21]:
train_df = pd.DataFrame(X_train_scaled, columns=X.columns)
train_df['target'] = y_train.values

val_df = pd.DataFrame(X_val_scaled, columns=X.columns)
val_df['target'] = y_val.values

test_df = pd.DataFrame(X_test_scaled, columns=X.columns)
test_df['target'] = y_test.values

# Save to CSV
os.makedirs("lab_1_csvs/", exist_ok=True)
train_df.to_csv('lab_1_csvs/train_scaled.csv', index=False)
val_df.to_csv('lab_1_csvs/validation_scaled.csv', index=False)
test_df.to_csv('lab_1_csvs/test_scaled.csv', index=False)


## Task 21 — Write a short data-quality summary

Write 5–7 bullet points summarising the dataset, class balance, missing values, duplicates, scaling need, and modelling risks.


In [22]:
summary = """
DATA QUALITY SUMMARY
===================

• Dataset Size: 569 samples with 30 numerical features extracted from breast cancer images

• Target Classes: Binary classification (0=malignant, 1=benign)

• Class Balance: Slightly imbalanced - 62.74% benign (357 samples), 37.26% malignant (212 samples)

• Missing Values: None - the dataset is complete with no missing values

• Duplicate Rows: None - no duplicate samples found

• Scaling Need: Critical - features have vastly different scales (range from 0.248 to 4254.0),
  requiring standardization for optimal model performance

• Potential Modelling Risks:
  - Multicollinearity: Several feature pairs have correlations >0.90, which may affect
    linear models and interpretability
  - Class Imbalance: The 63:37 split may bias models toward the majority class;
    consider using stratified sampling or class weights
  - Outliers: Several features contain outliers that may influence distance-based algorithms
  - Feature Selection: High correlations suggest dimensionality reduction could be beneficial
"""

print(summary)



DATA QUALITY SUMMARY

• Dataset Size: 569 samples with 30 numerical features extracted from breast cancer images

• Target Classes: Binary classification (0=malignant, 1=benign)

• Class Balance: Slightly imbalanced - 62.74% benign (357 samples), 37.26% malignant (212 samples)

• Missing Values: None - the dataset is complete with no missing values

• Duplicate Rows: None - no duplicate samples found

• Scaling Need: Critical - features have vastly different scales (range from 0.248 to 4254.0),
  requiring standardization for optimal model performance

• Potential Modelling Risks:
  - Multicollinearity: Several feature pairs have correlations >0.90, which may affect
    linear models and interpretability
  - Class Imbalance: The 63:37 split may bias models toward the majority class;
    consider using stratified sampling or class weights
  - Outliers: Several features contain outliers that may influence distance-based algorithms
  - Feature Selection: High correlations suggest dimensi

**Observation:** Write your interpretation here.